packages/marshal/src/types.js

```diff
@@ -2,70 +2,317 @@
+ * @typedef {{name: string]: T}} CopyRecord
+ * Corresponding PassStyle is "copyRecord".
  *
- * A Comparable is OnlyData when its pass-by-copy superstructure has no leaves,
- * i.e., when all the leaves of the data structure tree are primitive data
- * types or empty composites.
+ * The semantics of CopyRecords is modeled on Records from the Records and
+ * Tuples proposal.
+ * Property names can only be strings. A CopyRecord consists only of
+ * a set of (property-name, value) pairs, where each property name
+ * can appear at most once. It is therefore a single-valued mapping
+ * from property-name to value. CopyRecords enumerate their properties
+ * according to the sort order of their property names.
+ *
+ * CopyArray and CopyRecord are parameterized over various "deep" constraints
+ * explained below. For example, iff the values of a CopyRecord are
+ * PassableKeys, the CopyRecord is PassableKeys and can be compared for
+ * equivalence based on the pairwise equivalence of their values.
+ * Iff the values of a CopyRecord are sortable, then the CopyRecord is
+ * sortable. Two sortable CopyRecords with the same property names are
+ * compared by lexicographic comparison ordered by the sorted order of
+ * their property names.
  */
 
 /**
- * @typedef {OnlyData} PureData
+ * @typedef {Object} CopyError
+ * Corresponding PassStyle is "copyError".
+ *
+ * TODO What I'd like to say in these types in that only a CopyError
+ * without an `errorId` field is a Comparable.
  *
- * An OnlyData value is PureData when it contains no hidden mutable state,
- * e.g., when none of its pass-by-copy composite data objects are proxies. This
- * cannot be determined by inspection. It can only be achieved by trusted
- * construction. A PureData value cannot be used as a communications channel,
- * and can therefore be safely shared with subgraphs that should not be able
- * to communicate with each other.
+ * ***TODO WARNING*** Since CopyErrors have no analog amount Records and
+ * Tuples, perhaps a CopyError in general should only be a Comparable but
+ * not a PassableKey. Or perhaps not even a Comparable! In that case we
+ * wouldn't need to make a special case for `errorId`.
+ * @property {string} name
+ * @property {string} message
+ * @property {string=} errorId
  */
 
 /**
- * @typedef {*} Remotable
- * Might be an object explicitly deemed to be `Remotable`, an object inferred
- * to be Remotable, or a remote presence of a Remotable.
+ * @template {Passable} T
+ * @typedef { LeafData | CopyArray<T> | CopyRecord<T> | CopyError
+ * } NestedData
+ * Corresponding PassStyle values are
+ * the PassStyles of LeafData, "copyArray", "copyRecord", "copyError"
+ *
+ * NestedData is parameterized over various "deep" constraints. See
+ * the discussion in CopyArray and CopyRecord above.
  */
 
 /**
- * @typedef {Promise | Remotable} PassableCap
- * The leaves of a Passable's pass-by-copy superstructure.
+ * @typedef {NestedData<PureData>} PureData
```

Comment:
I appreciate the sentiment, but I believe that recursive types are impossible to do in JSDoc (they degrade just to `any`).  See `extra-types.d.ts` for where these types must be moved.

I'd be happy to be proven wrong, though!

Reply:
It is weird that the type system is so forgiving that when I say something meaningless it does not complain. I see in eslint-config/eslint-config.json the line
```json
    "jsdoc/valid-types": "off"
```
Is this why it is so silent about my screwups? What are the tradeoffs in turning this `"on"` instead?

Reply:
Attn @kriskowal @dckc

Reply:
I think `jsdoc/valid-types` checks type annotation syntax against jsdoc norms, not against typescript norms.

I don't think it's the type system that's so lax; it's the way tsc handles jsdoc. The only way I know to get precise feedback is to use .d.ts files.

Reply:
> Is this why it is so silent about my screwups? What are the tradeoffs in turning this `"on"` instead?

Yes, as #1751 says, that's why it's so soft and wishy-washy.  There is an upstream issue to enable more Typescript support in the JSDoc linter, but it needs to be almost completely finished before we'd be able to turn on the `jsdoc/valid-types` rule.